balloc/balloc.c

```diff
@@ -2251,6 +2252,30 @@ static int balloc_wild_scan_group(struct balloc_allocation_context *bac,
 	list = &grp->bgi_normal.bzp_extents;
 #endif
 
+	/**
+	 * Check to detect the block allocation request which came earlier
+	 * for criteria 1 on same group and come again for criteria 2 with
+	 * already set best extent.
+	 * In criteria 1 bac_status not marked to M0_BALLOC_AC_FOUND
+	 * because bac_found < M0_BALLOC_DEFAULT_MIN_TO_SCAN
+	 * Now with criteria 2 on same group with already set best extent may
+	 * not part of extent in free extent list because another requests
+	 * may have updated extents in free list.
+	 * Reset bac_found and best extent by detecting this case so that it
+	 * will find correct best extent and also set end_of_group flag so
+	 * balloc_check_limits() could call balloc_use_best_found() to set
+	 * final extent from best extent.
+	 */
+	if (bac->bac_found != 0) {
+		m0_bindex_t group = balloc_bn2gn(bac->bac_best.e_start,
+						 bac->bac_ctxt);
+		if (group == grp->bgi_groupno) {
+			end_of_group = 1;
+			bac->bac_found = 0;
```

Comment:
We should not set bac->bac_found = 0 as we do not want to reset this counter which would be helpful if we are not able to get any free blocks in this group because some other threads took them while we were still going through the other groups before we reached here.